### ISSUE TITLE
View is not removed after "remove" operation

### ISSUE THREAD

**Issue by un-tone** [Starcounter/Home#250]
I tested on WebsiteEditor app.

#### Steps to reproduce

1. Run **WebsiteEditor** and **WebsiteProvider** apps from `develop` branch.
1. Run **SignIn** app from `develop` branch.
1. If user is not created go to <http://localhost:8080/signin/createadminuser>.
1. Go to <http://localhost:8080/signin/signinuser> and log in.
1. Go to <http://localhost:8080/websiteeditor>.
1. Select "TopbarSurface" and go to "Routes" tab.
1. Create new route for the current page, e.g. `/WebsiteEditor/surface/G/routing` URI and unchecked **Allow route nesting** field.
1. Go to "Blending points" tab and expand "Main" item.
1. Create a new blending rule with `/signin/user` Blend URI and the created Route URI (`/WebsiteEditor/surface/G/routing`).
1. Go back to "Routes" tab and refresh the page.
1. Go to another tab, e.g. "General".

#### Expected result

`/signin/user` view does not appear on the "General" tab.

#### Actual result

`/signin/user` view still appears on the "General" tab despite it was removed from view model:

```js
{
    "op": "remove",
    "path": "/WebsiteProvider_0/Sections/Main/SignIn_0"
}, {
    "op": "remove",
    "path": "/WebsiteProvider_0/Sections/Main/SignIn_0"
}
```
Note that there are two same elements.

![image](https://user-images.githubusercontent.com/1625378/30830563-0c1f0d94-a24d-11e7-8e9b-c40283826781.png)

#### Notes

* It shows correct if I refresh the "General" page.

**Comment by warpech** [comment]
@alemoi can you pls take a look?

@un-tone is this a regression?

**Comment by warpech** [comment]
@alemoi can you pls take a look?

@un-tone is this a regression? Are any issues blocked by this?

@miyconst I advise this is included in 2.3.1

**Comment by alemoi** [comment]
@un-tone can you please check if wrong handlers are called? Just so we exclude the possibility that its a problem with Json patch?

**Comment by un-tone** [comment]
Yes, I checked. Wrong handlers are not called.

**Comment by alemoi** [comment]
> Yes, I checked. Wrong handlers are not called.

So what do you think, its a wrong Json patch coming then?

**Comment by un-tone** [comment]
It seems that JSON patch is correct, except of duplication of the "remove" operation item.

**Comment by alemoi** [comment]
But do you think that this duplication is caused by wrong handler call by Blender? Because as I understood:

> Wrong handlers are not called.

Meaning its not a Blender issue, but rather Json patch, or I'm probably wrong? Thanks for clarification!

**Comment by un-tone** [comment]
As it is shown on the screenshot the `/WebsiteProvider_0/Sections/Main/SignIn_0` element was removed from the View-model tree but is still shown in the UI. Please pay attention on it.

**Comment by warpech** [comment]
There are two issues here:

1. Client side `starcounter-include` does not unrender `SignIn_0` even though it is removed from the view-model. cc @tomalec
2. Server side Palindrom sends the removing patch twice. cc @chrhol

**Comment by warpech** [comment]
Re point 1: This line in `palindrom-polymer-client` is the problem: https://github.com/Palindrom/palindrom-polymer-client/blob/4.0.0/palindrom-client.html#L435. If the `undefined` in that line is changed to `null`, it works correctly.

I can see that on the master branch it is already fixed: https://github.com/Palindrom/palindrom-polymer-client/blob/ecb410b02df7ec52102d937ab535009929684453/palindrom-client.html#L435

@alshakero any comments on that? Is this safe to release in Starcounter 2.3.1?

**Comment by alshakero** [comment]
I don't think so. It's safer to branch from the released version, fix this issue. And release a patch. `palindrom-polymer-client`'s master is modified a lot and this would break a lot of stuff if paired with Polymer 1.x. 

Should I go for what I described?

**Comment by warpech** [comment]
We want to limit our engagement with Polymer 1.x (https://github.com/Starcounter/RebelsLounge/issues/66) as well as reduce the number of changes to Starcounter 2.3.1, but I think our templating engine is too important to skip fixing such issue. 

How much would it take to fix this and add tests for both branches (Polymer 1 and 2)?

**Comment by alshakero** [comment]
@warpech sorry for the late response. I can't predict how long it would take, maybe an hour if the change doesn't break tests with Polymer 1.x. It might take a while if they decided to break. I suggest trying the easy solution and giving up on it if it didn't work. WDTY?

**Comment by alshakero** [comment]
This is not tracked here https://github.com/Palindrom/palindrom-polymer-client/issues/47

**Comment by chrhol** [comment]
Is this fixed already, or is there some workaround in place in WebsiteEditor? 

Running the latest `develop` versions of website and signin, when I click on `General` as the last step, I get a patch that replaces the root, and everything seems to be fine. So no duplicate remove patches in my case.

**Comment by warpech** [comment]
@chrhol The problem is still reproducible for me. 

Are you running the three apps: WebsiteEditor, WebsiteProvider, SignIn in an empty database?

My versions: 
- Starcounter 2.3.1.7422
- https://github.com/StarcounterApps/Website/tree/8af81c32c94caf432eb3aad42063b4a221fd80ef
- https://github.com/StarcounterApps/SignIn/tree/75b5bfb206afe34874866011b642e6d842955890

**Comment by chrhol** [comment]
Weird, I must have done something wrong before, or I did not start from an empty database. Now I get the duplicated patches as well.

**Comment by chrhol** [comment]
This issue drives me crazy. I spend a ridiculous amount of time just reproducing it, divided between me messing up the instructions and the Starcounter version changing.

In version `2.3.1.7422` I can reproduce the duplicated remove patches. They come from (as usual) that the TypedJSON instance from signin is being used in two places. Now why the patch is duplicated is unclear, it shouldn't be. 

In the latest Starcounter version, following the instructions in the first comment doesn't seem to blend `SignIn` at all to the main-area for the `Routing` tab.

**Comment by un-tone** [comment]
> In the latest Starcounter version, following the instructions in the first comment doesn't seem to blend SignIn at all to the main-area for the Routing tab.

This is a known issue related to https://github.com/Starcounter/Website/issues/237 and discussed in https://github.com/Starcounter/Website/issues/238.

**Comment by un-tone** [comment]
The issue is not reproducible in SC 2.3.1.7454. There is a difference in JSON Patch, currently it returns "replace" operation for full section: `path: "/WebsiteProvider_0/Sections/Main"`.

**Comment by un-tone** [comment]
The current behaviour seems not good to me but it is another issue. Closing this one...
